src/libstd/thread/local.rs

```diff
@@ -275,7 +276,10 @@ impl<T: 'static> LocalKey<T> {
         // operations a little differently and make this safe to call.
         mem::replace(&mut *ptr, Some(value));
 
-        (*ptr).as_ref().unwrap()
+        match *ptr {
+            Some(ref x) => x,
+            None => intrinsics::unreachable(),
```

Comment:
I think this really deserves a comment indicating why it was done -- could you add one? That way we can avoid confusion or cleanup removing this without knowing what to check

Reply:
Sure thing!

Reply:
Also don't we have [unreachable_unchecked](https://doc.rust-lang.org/beta/std/hint/fn.unreachable_unchecked.html) so that we do not have to call an intrinsic directly?

Reply:
Sure!